### PR TITLE
feat(treesitter): playground improvements

### PR DESCRIPTION
**Pull request by lewis6991**
- Render node ranges as virtual text
- Set `filettype=query`. The virtual text is to avoid parsing errors.
- Make sure highlighted text is always in view.

### CODE REVIEW

**Comment by clason** [comment]
Make sure to test with language name toggled on (`I`).

**Comment by lewis6991** [comment]
@clason found out `show_tree()` doesn't work with bash scripts (because `bash != sh`) so added a slightly obscure change to fix that. Basically only make `get_parser` use the buffer filetype as a last resort.

**Comment by lewis6991** [comment]
That will work. But the buffer does have a parser stored, so if language is omitted we should just use that.

**Comment by clason** [comment]
Do we need the `lang` key then?

**Comment by lewis6991** [comment]
I would say no, but there is an very-edge case, where someone who want to use `show_tree` without highlighting enabled will need to be able to pass lang.

File: runtime/lua/vim/treesitter.lua [hunk 1]
```diff
@@ -462,6 +462,15 @@ function M.show_tree(opts)
         end_col = math.max(0, pos.end_col),
         hl_group = 'Visual',
       })
+
+      local topline, botline = vim.fn.line('w0', win), vim.fn.line('w0', win)
```
Comment (gpanders) [inline runtime/lua/vim/treesitter.lua:466]: Should botline be `w$`?

**Review by gpanders — COMMENTED** [review]
Code itself looks fine to me, though it might be a few days before I can actually test it. If other people have tested and approved then 👍 from me.

File: runtime/lua/vim/treesitter.lua (hunk 1 above)
Comment (lewis6991) [reply runtime/lua/vim/treesitter.lua:466]: Woops

**Review by gpanders — APPROVED** [review]
Just tested and looks good to me.

**Comment by clason** [comment]
@lewis6991 After using this a bit, I have to admit I'm not very happy about the virtual text changes, as this makes it much harder to yank the tree to copy it into Github issues (which is one of the main use cases for the playground). Can we revert this?

**Comment by max397574** [comment]
> Set filettype=query. The virtual text is to avoid parsing errors.

I think you have to choose what you want
being able to copy or being able to get highlighting

perhaps there would be a solution to allow both
e.g. with a `TextYankPost` (iirc that's how the even is called) autocommand to set the register contents to everything, including the virtual text
or map the `y` in the buffer
idk if that's something you want though

**Comment by lewis6991** [comment]
Could we control this with an argument to show tree? I'd really like to keep the syntax highlighting.

**Comment by clason** [comment]
Is there so much syntax highlighting in the tree? Can you show a screenshot?

**Comment by lewis6991** [comment]
> Does query have a comment string? Then we could just mark the annotations as comments.

Good idea.

**Comment by clason** [comment]
(you could still add the virtual text to keep the highlighting)

**Comment by lewis6991** [comment]
Isn't the virtual text the main complaint since that can't be copied?

**Comment by clason** [comment]
If it's _only_ virtual text, yes. Adding virtual text on top of actual text to have highlighted comments is the best of both worlds :)

**Comment by lewis6991** [comment]
I've implemented this without any virtual text:

image

Hows that?

**Comment by clason** [comment]
I'm particularly interested in the injected languages (moreso than the ranges).

**Comment by lewis6991** [comment]
That still works:

image

**Comment by lewis6991** [comment]
Let's continue in #22418

**Comment by clason** [comment]
Right, and I was wondering whether it's (easily) possible to keep the highlight for the language, as seeing that at a glance is often very helpful:
Screenshot 2023-02-26 at 14 00 39

(It's not a dealbreaker; as I said, if I had to choose between pasting and highlighting

**Pull request merged**